### PR TITLE
API + Permissions

### PR DESCRIPTION
**Pull request by alvinwan**
depends on #527 

- [x] #561 fix error propagation
- [x] #560 generalize permissions checks

ensure that *all* endpoints have permissions checks:
- [x] ParticipantAPI
- [x] UserAPI
- [x] AssignmentAPI
- [x] SubmissionAPI
- [x] SearchAPI
- [x] VersionAPI
- [x] CourseAPI
- [x] GroupAPI
- [x] QueueAPI
- [x] QueuesAPI
- [x] FinalSubmissionAPI
- [x] AnalyticsAPI

### CODE REVIEW

File: server/app/api.py [hunk 1]
```diff
@@ -1227,72 +1194,6 @@ def score(self, obj, user, data):
         subm.put()
         return score
 
-    def get_assignment(self, name):
```
Comment (Sumukh) [inline server/app/api.py]: Why was this and the submission method deleted?

Comment (alvinwan) [inline server/app/api.py]: Actually, it was just moved a couple dozen lines upward haha. I didn't remove it.

Comment (Sumukh) [inline server/app/api.py]: Ah thanks. I missed that big green chunk.

File: server/app/api.py [hunk 2]
```diff
@@ -1048,56 +975,114 @@ def data_for_zip(self, obj):
...
-        """
+        """Makes a zip response using a zip object."""
         response = make_response(zipfile)
         response.headers['Content-Disposition'] = (
             'attachment; filename=submission-%s.zip' % name)
         response.headers['Content-Type'] = 'application/zip'
         return response
 
+    def get_assignment(self, name):
+        """
```
Comment (alvinwan) [inline server/app/api.py]: @Sumukh Here it is.

File: server/app/api.py [hunk 3]
```diff
@@ -911,16 +855,7 @@ def lookup_assignments_by_name(self, name):
         return assignments
 
     def create_submission(self, user, assignment, messages, submit, submitter):
-        """
-        Create submission using user as parent to ensure ordering.
-
-        :param user: (object) caller
```
Comment (Sumukh) [inline server/app/api.py]: Can we keep this docstring? I think it's helpful

File: server/tests/integration/test_api_submission.py [hunk 4]
```diff
@@ -106,23 +106,24 @@ def test_zip(self):
...
 		""" Tests that zip_files does not crash """
 		name, file_contents = 'yolo', {'gup.py': 'import fish'}
 		name2, zipfile = self.API().zip_files(name, file_contents)
 		self.assertEqual(name, name2)
 		return name, zipfile
-		
+
 	def test_make_zip_response(self):
 		""" Check zipfile response headers """
 		with self.app.test_request_context('/api/v2'):
```
Comment (Sumukh) [inline server/tests/integration/test_api_submission.py]: @alvinwan - I might have missed something but why is this going to v2 of the api? Which API calls support v2?

Comment (alvinwan) [inline server/tests/integration/test_api_submission.py]: Not sure. I copied this from another section of the code. Do you think it's supposed to be v1?

**Comment by Sumukh** [comment]
Closing. I think we've migrated over a lot of the changes made here.

**Pull request closed without merging**